bors r- 

looks like merge skew 

```
# github.com/cockroachdb/cockroach/pkg/migration/migrations
pkg/migration/migrations/remove_invalid_database_privileges.go:92:71: cannot use idAndVersion.id (type catid.DescID) as type *kv.Txn in argument to descriptors.GetMutableDescriptorByID
pkg/migration/migrations/remove_invalid_database_privileges.go:92:71: cannot use txn (type *kv.Txn) as type catid.DescID in argument to descriptors.GetMutableDescriptorByID
```

--- a/pkg/bench/rttanalysis/testdata/benchmark_expectations
+++ b/pkg/bench/rttanalysis/testdata/benchmark_expectations
@@ -48,8 +48,8 @@ exp,benchmark
 29-32,DropView/drop_2_views
 36,DropView/drop_3_views
 19,Grant/grant_all_on_1_table
-21-22,Grant/grant_all_on_2_tables
-23-24,Grant/grant_all_on_3_tables
+19,Grant/grant_all_on_2_tables
+19,Grant/grant_all_on_3_tables
 17,GrantRole/grant_1_role
 20,GrantRole/grant_2_roles
 2,ORMQueries/activerecord_type_introspection_query
@@ -70,8 +70,8 @@ exp,benchmark
 4,ORMQueries/pg_namespace
 2,ORMQueries/pg_type
 19,Revoke/revoke_all_on_1_table
-21,Revoke/revoke_all_on_2_tables
-23,Revoke/revoke_all_on_3_tables
+19,Revoke/revoke_all_on_2_tables
+19,Revoke/revoke_all_on_3_tables
 16,RevokeRole/revoke_1_role
 18,RevokeRole/revoke_2_roles
 1,SystemDatabaseQueries/select_system.users_with_empty_database_Name

--- a/pkg/ccl/backupccl/restore_job.go
+++ b/pkg/ccl/backupccl/restore_job.go
@@ -1089,7 +1089,7 @@ func createImportingDescriptors(
 			// Write the updated databases.
 			for dbID, schemas := range existingDBsWithNewSchemas {
 				log.Infof(ctx, "writing %d schema entries to database %d", len(schemas), dbID)
-				desc, err := descsCol.GetMutableDescriptorByID(ctx, dbID, txn)
+				desc, err := descsCol.GetMutableDescriptorByID(ctx, txn, dbID)
 				if err != nil {
 					return err
 				}
@@ -1924,7 +1924,7 @@ func (r *restoreResumer) publishDescriptors(
 		}
 	}
 	for _, sc := range details.SchemaDescs {
-		mutDesc, err := descsCol.GetMutableDescriptorByID(ctx, sc.ID, txn)
+		mutDesc, err := descsCol.GetMutableDescriptorByID(ctx, txn, sc.ID)
 		if err != nil {
 			return err
 		}
@@ -1940,7 +1940,7 @@ func (r *restoreResumer) publishDescriptors(
 		// an offline state.
 		// TODO(lucy): Should we make this more explicit with a format version
 		// field in the details?
-		mutDesc, err := descsCol.GetMutableDescriptorByID(ctx, dbDesc.ID, txn)
+		mutDesc, err := descsCol.GetMutableDescriptorByID(ctx, txn, dbDesc.ID)
 		if err != nil {
 			return err
 		}
@@ -2216,7 +2216,7 @@ func (r *restoreResumer) dropDescriptors(
 			continue
 		}
 
-		mutSchema, err := descsCol.GetMutableDescriptorByID(ctx, schemaDesc.GetID(), txn)
+		mutSchema, err := descsCol.GetMutableDescriptorByID(ctx, txn, schemaDesc.GetID())
 		if err != nil {
 			return err
 		}
@@ -2244,7 +2244,7 @@ func (r *restoreResumer) dropDescriptors(
 	// modified schema slice after we have issued a `b.Del` to drop it.
 	for dbID, schemas := range dbsWithDeletedSchemas {
 		log.Infof(ctx, "deleting %d schema entries from database %d", len(schemas), dbID)
-		desc, err := descsCol.GetMutableDescriptorByID(ctx, dbID, txn)
+		desc, err := descsCol.GetMutableDescriptorByID(ctx, txn, dbID)
 		if err != nil {
 			return err
 		}
@@ -2286,7 +2286,7 @@ func (r *restoreResumer) dropDescriptors(
 			continue
 		}
 
-		db, err := descsCol.GetMutableDescriptorByID(ctx, dbDesc.GetID(), txn)
+		db, err := descsCol.GetMutableDescriptorByID(ctx, txn, dbDesc.GetID())
 		if err != nil {
 			return err
 		}

--- a/pkg/ccl/importccl/import_job.go
+++ b/pkg/ccl/importccl/import_job.go
@@ -588,7 +588,7 @@ func (r *importResumer) prepareSchemasForIngestion(
 	schemaMetadata.schemaPreparedDetails.Schemas = make([]jobspb.ImportDetails_Schema,
 		len(details.Schemas))
 
-	desc, err := descsCol.GetMutableDescriptorByID(ctx, details.ParentID, txn)
+	desc, err := descsCol.GetMutableDescriptorByID(ctx, txn, details.ParentID)
 	if err != nil {
 		return nil, err
 	}
@@ -1042,7 +1042,7 @@ func (r *importResumer) publishSchemas(ctx context.Context, execCfg *sql.Executo
 	) error {
 		b := txn.NewBatch()
 		for _, schema := range details.Schemas {
-			newDesc, err := descsCol.GetMutableDescriptorByID(ctx, schema.Desc.GetID(), txn)
+			newDesc, err := descsCol.GetMutableDescriptorByID(ctx, txn, schema.Desc.GetID())
 			if err != nil {
 				return err
 			}
@@ -1515,7 +1515,7 @@ func (r *importResumer) dropSchemas(
 	}
 
 	// Resolve the database descriptor.
-	desc, err := descsCol.GetMutableDescriptorByID(ctx, details.ParentID, txn)
+	desc, err := descsCol.GetMutableDescriptorByID(ctx, txn, details.ParentID)
 	if err != nil {
 		return nil, err
 	}
@@ -1528,7 +1528,7 @@ func (r *importResumer) dropSchemas(
 
 	droppedSchemaIDs := make([]descpb.ID, 0)
 	for _, schema := range details.Schemas {
-		desc, err := descsCol.GetMutableDescriptorByID(ctx, schema.Desc.ID, txn)
+		desc, err := descsCol.GetMutableDescriptorByID(ctx, txn, schema.Desc.ID)
 		if err != nil {
 			return nil, err
 		}

--- a/pkg/migration/migrations/remove_invalid_database_privileges.go
+++ b/pkg/migration/migrations/remove_invalid_database_privileges.go
@@ -89,7 +89,7 @@ func fixDescriptors(
 		for _, idAndVersion := range descriptorIDAndVersions {
 			// GetMutableDescriptorByID calls RunPostDeserializationChanges which
 			// fixes the descriptor.
-			desc, err := descriptors.GetMutableDescriptorByID(ctx, idAndVersion.id, txn)
+			desc, err := descriptors.GetMutableDescriptorByID(ctx, txn, idAndVersion.id)
 			if err != nil {
 				return err
 			}

--- a/pkg/sql/alter_schema.go
+++ b/pkg/sql/alter_schema.go
@@ -146,7 +146,7 @@ func (p *planner) alterSchemaOwner(
 	}
 
 	// The user must also have CREATE privilege on the schema's database.
-	parentDBDesc, err := p.Descriptors().GetMutableDescriptorByID(ctx, scDesc.GetParentID(), p.txn)
+	parentDBDesc, err := p.Descriptors().GetMutableDescriptorByID(ctx, p.txn, scDesc.GetParentID())
 	if err != nil {
 		return err
 	}

--- a/pkg/sql/catalog/descs/collection_test.go
+++ b/pkg/sql/catalog/descs/collection_test.go
@@ -226,7 +226,7 @@ func TestAddUncommittedDescriptorAndMutableResolution(t *testing.T) {
 
 			require.Same(t, db, resolved)
 
-			byID, err := descriptors.GetMutableDescriptorByID(ctx, db.GetID(), txn)
+			byID, err := descriptors.GetMutableDescriptorByID(ctx, txn, db.GetID())
 			require.NoError(t, err)
 			require.Same(t, db, byID)
 
@@ -300,7 +300,7 @@ func TestAddUncommittedDescriptorAndMutableResolution(t *testing.T) {
 
 			require.Same(t, schema, resolved)
 
-			byID, err := descriptors.GetMutableDescriptorByID(ctx, schema.GetID(), txn)
+			byID, err := descriptors.GetMutableDescriptorByID(ctx, txn, schema.GetID())
 			require.NoError(t, err)
 
 			require.Same(t, schema, byID)
@@ -324,7 +324,7 @@ func TestAddUncommittedDescriptorAndMutableResolution(t *testing.T) {
 
 			require.Same(t, tab, resolved)
 
-			byID, err := descriptors.GetMutableDescriptorByID(ctx, tab.GetID(), txn)
+			byID, err := descriptors.GetMutableDescriptorByID(ctx, txn, tab.GetID())
 			require.NoError(t, err)
 
 			require.Same(t, tab, byID)

--- a/pkg/sql/catalog/descs/database.go
+++ b/pkg/sql/catalog/descs/database.go
@@ -96,7 +96,7 @@ func (tc *Collection) GetImmutableDatabaseByID(
 func (tc *Collection) getDatabaseByID(
 	ctx context.Context, txn *kv.Txn, dbID descpb.ID, flags tree.DatabaseLookupFlags,
 ) (bool, catalog.DatabaseDescriptor, error) {
-	desc, err := tc.getDescriptorByID(ctx, txn, dbID, flags)
+	descs, err := tc.getDescriptorsByID(ctx, txn, flags, dbID)
 	if err != nil {
 		if errors.Is(err, catalog.ErrDescriptorNotFound) {
 			if flags.Required {
@@ -106,7 +106,7 @@ func (tc *Collection) getDatabaseByID(
 		}
 		return false, nil, err
 	}
-	db, ok := desc.(catalog.DatabaseDescriptor)
+	db, ok := descs[0].(catalog.DatabaseDescriptor)
 	if !ok {
 		return false, nil, sqlerrors.NewUndefinedDatabaseError(fmt.Sprintf("[%d]", dbID))
 	}

--- a/pkg/sql/catalog/descs/descriptor.go
+++ b/pkg/sql/catalog/descs/descriptor.go
@@ -28,25 +28,16 @@ import (
 
 // GetMutableDescriptorByID returns a mutable implementation of the descriptor
 // with the requested id. An error is returned if no descriptor exists.
-// Deprecated in favor of GetMutableDescriptorByIDWithFlags.
 func (tc *Collection) GetMutableDescriptorByID(
-	ctx context.Context, id descpb.ID, txn *kv.Txn,
-) (catalog.MutableDescriptor, error) {
-	return tc.GetMutableDescriptorByIDWithFlags(ctx, txn, id, tree.CommonLookupFlags{
-		IncludeOffline: true,
-		IncludeDropped: true,
-	})
-}
-
-// GetMutableDescriptorByIDWithFlags returns a mutable implementation of the
-// descriptor with the requested id. An error is returned if no descriptor exists.
-// TODO (lucy): This is meant to replace GetMutableDescriptorByID. Once it does,
-// rename this function.
-func (tc *Collection) GetMutableDescriptorByIDWithFlags(
-	ctx context.Context, txn *kv.Txn, id descpb.ID, flags tree.CommonLookupFlags,
+	ctx context.Context, txn *kv.Txn, id descpb.ID,
 ) (catalog.MutableDescriptor, error) {
 	log.VEventf(ctx, 2, "planner getting mutable descriptor for id %d", id)
-	flags.RequireMutable = true
+	flags := tree.CommonLookupFlags{
+		Required:       true,
+		RequireMutable: true,
+		IncludeOffline: true,
+		IncludeDropped: true,
+	}
 	desc, err := tc.getDescriptorByID(ctx, txn, id, flags)
 	if err != nil {
 		return nil, err

--- a/pkg/sql/catalog/descs/descriptor.go
+++ b/pkg/sql/catalog/descs/descriptor.go
@@ -26,122 +26,215 @@ import (
 	"github.com/cockroachdb/errors"
 )
 
-// GetMutableDescriptorByID returns a mutable implementation of the descriptor
-// with the requested id. An error is returned if no descriptor exists.
-func (tc *Collection) GetMutableDescriptorByID(
-	ctx context.Context, txn *kv.Txn, id descpb.ID,
-) (catalog.MutableDescriptor, error) {
-	log.VEventf(ctx, 2, "planner getting mutable descriptor for id %d", id)
+// GetMutableDescriptorsByID returns a mutable implementation of the descriptors
+// with the requested ids. An error is returned if no descriptor exists.
+func (tc *Collection) GetMutableDescriptorsByID(
+	ctx context.Context, txn *kv.Txn, ids ...descpb.ID,
+) ([]catalog.MutableDescriptor, error) {
 	flags := tree.CommonLookupFlags{
 		Required:       true,
 		RequireMutable: true,
 		IncludeOffline: true,
 		IncludeDropped: true,
 	}
-	desc, err := tc.getDescriptorByID(ctx, txn, id, flags)
+	descs, err := tc.getDescriptorsByID(ctx, txn, flags, ids...)
 	if err != nil {
 		return nil, err
 	}
-	return desc.(catalog.MutableDescriptor), nil
+	ret := make([]catalog.MutableDescriptor, len(descs))
+	for i, desc := range descs {
+		ret[i] = desc.(catalog.MutableDescriptor)
+	}
+	return ret, nil
 }
 
-// GetImmutableDescriptorByID returns an immmutable implementation of the
-// descriptor with the requested id. An error is returned if no descriptor exists.
-// Deprecated in favor of GetMutableDescriptorByIDWithFlags.
+// GetMutableDescriptorByID delegates to GetMutableDescriptorsByID.
+func (tc *Collection) GetMutableDescriptorByID(
+	ctx context.Context, txn *kv.Txn, id descpb.ID,
+) (catalog.MutableDescriptor, error) {
+	descs, err := tc.GetMutableDescriptorsByID(ctx, txn, id)
+	if err != nil {
+		return nil, err
+	}
+	return descs[0], nil
+}
+
+// GetImmutableDescriptorsByID returns an immutable implementation of the
+// descriptors with the requested ids. An error is returned if no descriptor
+// exists, regardless of whether the Required flag is set or not.
+func (tc *Collection) GetImmutableDescriptorsByID(
+	ctx context.Context, txn *kv.Txn, flags tree.CommonLookupFlags, ids ...descpb.ID,
+) ([]catalog.Descriptor, error) {
+	flags.RequireMutable = false
+	return tc.getDescriptorsByID(ctx, txn, flags, ids...)
+}
+
+// GetImmutableDescriptorByID delegates to GetImmutableDescriptorsByID.
 func (tc *Collection) GetImmutableDescriptorByID(
 	ctx context.Context, txn *kv.Txn, id descpb.ID, flags tree.CommonLookupFlags,
 ) (catalog.Descriptor, error) {
-	log.VEventf(ctx, 2, "planner getting immutable descriptor for id %d", id)
-	flags.RequireMutable = false
-	return tc.getDescriptorByID(ctx, txn, id, flags)
-}
-
-func (tc *Collection) getDescriptorByID(
-	ctx context.Context, txn *kv.Txn, id descpb.ID, flags tree.CommonLookupFlags,
-) (catalog.Descriptor, error) {
-	return tc.getDescriptorByIDMaybeSetTxnDeadline(
-		ctx, txn, id, flags, false /* setTxnDeadline */)
-}
-
-// getDescriptorByIDMaybeSetTxnDeadline returns a descriptor according to the
-// provided lookup flags. Note that flags.Required is ignored, and an error is
-// always returned if no descriptor with the ID exists.
-func (tc *Collection) getDescriptorByIDMaybeSetTxnDeadline(
-	ctx context.Context, txn *kv.Txn, id descpb.ID, flags tree.CommonLookupFlags, setTxnDeadline bool,
-) (catalog.Descriptor, error) {
-	getDescriptorByID := func() (catalog.Descriptor, error) {
-		vd, err := tc.virtual.getByID(ctx, id, flags.RequireMutable)
-		if vd != nil || err != nil {
-			return vd, err
-		}
-
-		if found, sd := tc.synthetic.getByID(id); found && !flags.AvoidSynthetic {
-			if flags.RequireMutable {
-				return nil, newMutableSyntheticDescriptorAssertionError(sd.GetID())
-			}
-			return sd, nil
-		}
-
-		{
-			ud := tc.uncommitted.getByID(id)
-			if ud != nil {
-				log.VEventf(ctx, 2, "found uncommitted descriptor %d", id)
-				if flags.RequireMutable {
-					ud, err = tc.uncommitted.checkOut(id)
-					if err != nil {
-						return nil, err
-					}
-				}
-				return ud, nil
-			}
-		}
-
-		if !flags.AvoidLeased && !flags.RequireMutable && !lease.TestingTableLeasesAreDisabled() {
-			// If we have already read all of the descriptors, use it as a negative
-			// cache to short-circuit a lookup we know will be doomed to fail.
-			//
-			// TODO(ajwerner): More generally leverage this set of kv descriptors on
-			// the resolution path.
-			if tc.kv.idDefinitelyDoesNotExist(id) {
-				return nil, catalog.ErrDescriptorNotFound
-			}
-
-			desc, shouldReadFromStore, err := tc.leased.getByID(ctx, tc.deadlineHolder(txn), id, setTxnDeadline)
-			if err != nil {
-				return nil, err
-			}
-			if !shouldReadFromStore {
-				return desc, nil
-			}
-		}
-
-		return tc.withReadFromStore(flags.RequireMutable, func() (catalog.MutableDescriptor, error) {
-			return tc.kv.getByID(ctx, txn, id)
-		})
-	}
-
-	desc, err := getDescriptorByID()
+	descs, err := tc.GetImmutableDescriptorsByID(ctx, txn, flags, id)
 	if err != nil {
 		return nil, err
 	}
-	if dropped, err := filterDescriptorState(desc, true /* required */, flags); err != nil || dropped {
-		// This is a special case for tables in the adding state: Roughly speaking,
-		// we always need to resolve tables in the adding state by ID when they were
-		// newly created in the transaction for DDL statements and for some
-		// information queries (but not for ordinary name resolution for queries/
-		// DML), but we also need to make these tables public in the schema change
-		// job in a separate transaction.
-		// TODO (lucy): We need something like an IncludeAdding flag so that callers
-		// can specify this behavior, instead of having the collection infer the
-		// desired behavior based on the flags (and likely producing unintended
-		// behavior). See the similar comment on etDescriptorByName, which covers
-		// the ordinary name resolution path as well as DDL statements.
-		if desc.Adding() && (desc.IsUncommittedVersion() || flags.AvoidLeased || flags.RequireMutable) {
-			return desc, nil
+	return descs[0], nil
+}
+
+// getDescriptorsByID returns a slice of descriptors by ID according to the
+// provided lookup flags. Note that flags.Required is ignored, and an error is
+// always returned if no descriptor with the ID exists.
+func (tc *Collection) getDescriptorsByID(
+	ctx context.Context, txn *kv.Txn, flags tree.CommonLookupFlags, ids ...descpb.ID,
+) (descs []catalog.Descriptor, err error) {
+	defer func() {
+		if err == nil {
+			err = filterDescriptorsStates(descs, flags)
 		}
+		if err != nil {
+			descs = nil
+		}
+	}()
+
+	log.VEventf(ctx, 2, "looking up descriptors for ids %d", ids)
+	descs = make([]catalog.Descriptor, len(ids))
+	{
+		// Look up the descriptors in all layers except the KV layer on a
+		// best-effort basis.
+		q := byIDLookupContext{
+			ctx:   ctx,
+			txn:   txn,
+			tc:    tc,
+			flags: flags,
+		}
+		for _, fn := range []func(id descpb.ID) (catalog.Descriptor, error){
+			q.lookupVirtual,
+			q.lookupSynthetic,
+			q.lookupUncommitted,
+			q.lookupLeased,
+		} {
+			for i, id := range ids {
+				if descs[i] != nil {
+					continue
+				}
+				desc, err := fn(id)
+				if err != nil {
+					return nil, err
+				}
+				descs[i] = desc
+			}
+		}
+	}
+
+	kvIDs := make([]descpb.ID, 0, len(ids))
+	indexes := make([]int, 0, len(ids))
+	for i, id := range ids {
+		if descs[i] != nil {
+			continue
+		}
+		kvIDs = append(kvIDs, id)
+		indexes = append(indexes, i)
+	}
+	if len(kvIDs) == 0 {
+		// No KV lookup necessary, return early.
+		return descs, nil
+	}
+	kvDescs, err := tc.withReadFromStore(flags.RequireMutable, func() ([]catalog.MutableDescriptor, error) {
+		return tc.kv.getByIDs(ctx, txn, kvIDs)
+	})
+	if err != nil {
+		return nil, err
+	}
+	for j, desc := range kvDescs {
+		descs[indexes[j]] = desc
+	}
+	return descs, nil
+}
+
+// byIDLookupContext is a helper struct for getDescriptorsByID which contains
+// the parameters for looking up descriptors by ID at various levels in the
+// Collection.
+type byIDLookupContext struct {
+	ctx   context.Context
+	txn   *kv.Txn
+	tc    *Collection
+	flags tree.CommonLookupFlags
+}
+
+func (q *byIDLookupContext) lookupVirtual(id descpb.ID) (catalog.Descriptor, error) {
+	return q.tc.virtual.getByID(q.ctx, id, q.flags.RequireMutable)
+}
+
+func (q *byIDLookupContext) lookupSynthetic(id descpb.ID) (catalog.Descriptor, error) {
+	if q.flags.AvoidSynthetic {
+		return nil, nil
+	}
+	_, sd := q.tc.synthetic.getByID(id)
+	if sd == nil {
+		return nil, nil
+	}
+	if q.flags.RequireMutable {
+		return nil, newMutableSyntheticDescriptorAssertionError(sd.GetID())
+	}
+	return sd, nil
+}
+
+func (q *byIDLookupContext) lookupUncommitted(id descpb.ID) (_ catalog.Descriptor, err error) {
+	ud := q.tc.uncommitted.getByID(id)
+	if ud == nil {
+		return nil, nil
+	}
+	log.VEventf(q.ctx, 2, "found uncommitted descriptor %d", id)
+	if !q.flags.RequireMutable {
+		return ud, nil
+	}
+	return q.tc.uncommitted.checkOut(id)
+}
+
+func (q *byIDLookupContext) lookupLeased(id descpb.ID) (catalog.Descriptor, error) {
+	if q.flags.AvoidLeased || q.flags.RequireMutable || lease.TestingTableLeasesAreDisabled() {
+		return nil, nil
+	}
+	// If we have already read all of the descriptors, use it as a negative
+	// cache to short-circuit a lookup we know will be doomed to fail.
+	//
+	// TODO(ajwerner): More generally leverage this set of kv descriptors on
+	// the resolution path.
+	if q.tc.kv.idDefinitelyDoesNotExist(id) {
+		return nil, catalog.ErrDescriptorNotFound
+	}
+	desc, shouldReadFromStore, err := q.tc.leased.getByID(q.ctx, q.tc.deadlineHolder(q.txn), id)
+	if err != nil || shouldReadFromStore {
 		return nil, err
 	}
 	return desc, nil
+}
+
+// filterDescriptorsStates is a helper function for getDescriptorsByID.
+func filterDescriptorsStates(descs []catalog.Descriptor, flags tree.CommonLookupFlags) error {
+	for _, desc := range descs {
+		// The first return value can safely be ignored, it will always be false
+		// because the required flag is set.
+		_, err := filterDescriptorState(desc, true /* required */, flags)
+		if err == nil {
+			continue
+		}
+		if desc.Adding() && (desc.IsUncommittedVersion() || flags.AvoidLeased || flags.RequireMutable) {
+			// This is a special case for tables in the adding state: Roughly speaking,
+			// we always need to resolve tables in the adding state by ID when they were
+			// newly created in the transaction for DDL statements and for some
+			// information queries (but not for ordinary name resolution for queries/
+			// DML), but we also need to make these tables public in the schema change
+			// job in a separate transaction.
+			// TODO (lucy): We need something like an IncludeAdding flag so that callers
+			// can specify this behavior, instead of having the collection infer the
+			// desired behavior based on the flags (and likely producing unintended
+			// behavior). See the similar comment on etDescriptorByName, which covers
+			// the ordinary name resolution path as well as DDL statements.
+			continue
+		}
+		return err
+	}
+	return nil
 }
 
 func (tc *Collection) getByName(
@@ -197,11 +290,19 @@ func (tc *Collection) getByName(
 		}
 	}
 
-	desc, err = tc.withReadFromStore(mutable, func() (desc catalog.MutableDescriptor, err error) {
+	var descs []catalog.Descriptor
+	descs, err = tc.withReadFromStore(mutable, func() ([]catalog.MutableDescriptor, error) {
 		uncommittedDB, _ := tc.uncommitted.getByID(parentID).(catalog.DatabaseDescriptor)
-		return tc.kv.getByName(ctx, txn, uncommittedDB, parentID, parentSchemaID, name)
+		desc, err := tc.kv.getByName(ctx, txn, uncommittedDB, parentID, parentSchemaID, name)
+		if err != nil {
+			return nil, err
+		}
+		return []catalog.MutableDescriptor{desc}, nil
 	})
-	return desc != nil, desc, err
+	if err != nil {
+		return false, nil, err
+	}
+	return true, descs[0], err
 }
 
 // withReadFromStore updates the state of the Collection, especially its
@@ -209,24 +310,31 @@ func (tc *Collection) getByName(
 // layer. The logic is the same regardless of whether the descriptor was read
 // by name or by ID.
 func (tc *Collection) withReadFromStore(
-	requireMutable bool, readFn func() (catalog.MutableDescriptor, error),
-) (desc catalog.Descriptor, _ error) {
-	mut, err := readFn()
-	if mut == nil || err != nil {
-		return nil, err
-	}
-	desc, err = tc.uncommitted.add(mut)
+	requireMutable bool, readFn func() ([]catalog.MutableDescriptor, error),
+) (descs []catalog.Descriptor, _ error) {
+	muts, err := readFn()
 	if err != nil {
 		return nil, err
 	}
-	if requireMutable {
-		desc, err = tc.uncommitted.checkOut(desc.GetID())
+	descs = make([]catalog.Descriptor, len(muts))
+	for i, mut := range muts {
+		if mut == nil {
+			continue
+		}
+		desc, err := tc.uncommitted.add(mut)
 		if err != nil {
 			return nil, err
 		}
+		if requireMutable {
+			desc, err = tc.uncommitted.checkOut(desc.GetID())
+			if err != nil {
+				return nil, err
+			}
+		}
+		descs[i] = desc
 	}
 	tc.kv.releaseAllDescriptors()
-	return desc, nil
+	return descs, nil
 }
 
 func (tc *Collection) deadlineHolder(txn *kv.Txn) deadlineHolder {

--- a/pkg/sql/catalog/descs/dist_sql_type_resolver.go
+++ b/pkg/sql/catalog/descs/dist_sql_type_resolver.go
@@ -69,14 +69,12 @@ func (dt *DistSQLTypeResolver) GetTypeDescriptor(
 	flags := tree.CommonLookupFlags{
 		Required: true,
 	}
-	desc, err := dt.descriptors.getDescriptorByIDMaybeSetTxnDeadline(
-		ctx, dt.txn, id, flags, false, /* setTxnDeadline */
-	)
+	descs, err := dt.descriptors.getDescriptorsByID(ctx, dt.txn, flags, id)
 	if err != nil {
 		return tree.TypeName{}, nil, err
 	}
 	var typeDesc catalog.TypeDescriptor
-	switch t := desc.(type) {
+	switch t := descs[0].(type) {
 	case catalog.TypeDescriptor:
 		// User-defined type.
 		typeDesc = t
@@ -95,9 +93,9 @@ func (dt *DistSQLTypeResolver) GetTypeDescriptor(
 		}
 	default:
 		return tree.TypeName{}, nil, pgerror.Newf(pgcode.WrongObjectType,
-			"descriptor %d is a %s not a %s", id, desc.DescriptorType(), catalog.Type)
+			"descriptor %d is a %s not a %s", id, t.DescriptorType(), catalog.Type)
 	}
-	name := tree.MakeUnqualifiedTypeName(desc.GetName())
+	name := tree.MakeUnqualifiedTypeName(typeDesc.GetName())
 	return name, typeDesc, nil
 }
 

--- a/pkg/sql/catalog/descs/hydrate.go
+++ b/pkg/sql/catalog/descs/hydrate.go
@@ -49,7 +49,7 @@ func (tc *Collection) hydrateTypesInTableDesc(
 			if err != nil {
 				return tree.TypeName{}, nil, err
 			}
-			dbDesc, err := tc.GetMutableDescriptorByID(ctx, desc.ParentID, txn)
+			dbDesc, err := tc.GetMutableDescriptorByID(ctx, txn, desc.ParentID)
 			if err != nil {
 				return tree.TypeName{}, nil, err
 			}

--- a/pkg/sql/catalog/descs/kv_descriptors.go
+++ b/pkg/sql/catalog/descs/kv_descriptors.go
@@ -166,12 +166,12 @@ func (kd *kvDescriptors) getByName(
 	parentID descpb.ID,
 	parentSchemaID descpb.ID,
 	name string,
-) (desc catalog.MutableDescriptor, err error) {
+) (catalog.MutableDescriptor, error) {
 	descID, err := kd.lookupName(ctx, txn, maybeDB, parentID, parentSchemaID, name)
 	if err != nil || descID == descpb.InvalidID {
 		return nil, err
 	}
-	desc, err = kd.getByID(ctx, txn, descID)
+	descs, err := kd.getByIDs(ctx, txn, []descpb.ID{descID})
 	if err != nil {
 		if errors.Is(err, catalog.ErrDescriptorNotFound) {
 			// Having done the namespace lookup, the descriptor must exist.
@@ -179,7 +179,7 @@ func (kd *kvDescriptors) getByName(
 		}
 		return nil, err
 	}
-	if desc.GetName() != name {
+	if descs[0].GetName() != name {
 		// Immediately after a RENAME an old name still points to the descriptor
 		// during the drain phase for the name. Do not return a descriptor during
 		// draining.
@@ -190,31 +190,44 @@ func (kd *kvDescriptors) getByName(
 		// uncommitted namespace operations.
 		return nil, nil
 	}
-	return desc, nil
+	return descs[0], nil
 }
 
-// getByID actually reads a descriptor from the storage layer.
-func (kd *kvDescriptors) getByID(
-	ctx context.Context, txn *kv.Txn, id descpb.ID,
-) (_ catalog.MutableDescriptor, _ error) {
-	if id == keys.SystemDatabaseID {
-		// Special handling for the system database descriptor.
-		//
-		// This is done for performance reasons, to save ourselves an unnecessary
-		// round trip to storage which otherwise quickly compounds.
-		//
-		// The system database descriptor should never actually be mutated, which is
-		// why we return the same hard-coded descriptor every time. It's assumed
-		// that callers of this method will check the privileges on the descriptor
-		// (like any other database) and return an error.
-		return dbdesc.NewBuilder(systemschema.SystemDB.DatabaseDesc()).BuildExistingMutable(), nil
+// getByIDs actually reads a batch of descriptors from the storage layer.
+func (kd *kvDescriptors) getByIDs(
+	ctx context.Context, txn *kv.Txn, ids []descpb.ID,
+) ([]catalog.MutableDescriptor, error) {
+	ret := make([]catalog.MutableDescriptor, len(ids))
+	kvIDs := make([]descpb.ID, 0, len(ids))
+	indexes := make([]int, 0, len(ids))
+	for i, id := range ids {
+		if id == keys.SystemDatabaseID {
+			// Special handling for the system database descriptor.
+			//
+			// This is done for performance reasons, to save ourselves an unnecessary
+			// round trip to storage which otherwise quickly compounds.
+			//
+			// The system database descriptor should never actually be mutated, which is
+			// why we return the same hard-coded descriptor every time. It's assumed
+			// that callers of this method will check the privileges on the descriptor
+			// (like any other database) and return an error.
+			ret[i] = dbdesc.NewBuilder(systemschema.SystemDB.DatabaseDesc()).BuildExistingMutable()
+		} else {
+			kvIDs = append(kvIDs, id)
+			indexes = append(indexes, i)
+		}
 	}
-
-	imm, err := catkv.MustGetDescriptorByID(ctx, txn, kd.codec, id, catalog.Any)
+	if len(kvIDs) == 0 {
+		return ret, nil
+	}
+	kvDescs, err := catkv.MustGetDescriptorsByID(ctx, txn, kd.codec, kvIDs, catalog.Any)
 	if err != nil {
 		return nil, err
 	}
-	return imm.NewBuilder().BuildExistingMutable(), err
+	for j, desc := range kvDescs {
+		ret[indexes[j]] = desc.NewBuilder().BuildExistingMutable()
+	}
+	return ret, nil
 }
 
 func (kd *kvDescriptors) getAllDescriptors(

--- a/pkg/sql/catalog/descs/leased_descriptors.go
+++ b/pkg/sql/catalog/descs/leased_descriptors.go
@@ -110,10 +110,8 @@ func (ld *leasedDescriptors) getByName(
 
 // getByID return a leased descriptor valid for the transaction,
 // acquiring one if necessary.
-// We set a deadline on the transaction based on the lease expiration, which is
-// the usual case, unless setTxnDeadline is false.
 func (ld *leasedDescriptors) getByID(
-	ctx context.Context, txn deadlineHolder, id descpb.ID, setTxnDeadline bool,
+	ctx context.Context, txn deadlineHolder, id descpb.ID,
 ) (_ catalog.Descriptor, shouldReadFromStore bool, _ error) {
 	// First, look to see if we already have the table in the shared cache.
 	if cached := ld.getCachedByID(ctx, id); cached != nil {
@@ -126,6 +124,7 @@ func (ld *leasedDescriptors) getByID(
 
 	readTimestamp := txn.ReadTimestamp()
 	desc, err := ld.lm.Acquire(ctx, readTimestamp, id)
+	const setTxnDeadline = false
 	return ld.getResult(ctx, txn, setTxnDeadline, desc, err)
 }
 

--- a/pkg/sql/catalog/descs/schema.go
+++ b/pkg/sql/catalog/descs/schema.go
@@ -147,7 +147,7 @@ func (tc *Collection) getSchemaByID(
 	}
 
 	// Otherwise, fall back to looking up the descriptor with the desired ID.
-	desc, err := tc.getDescriptorByID(ctx, txn, schemaID, flags)
+	descs, err := tc.getDescriptorsByID(ctx, txn, flags, schemaID)
 	if err != nil {
 		if errors.Is(err, catalog.ErrDescriptorNotFound) {
 			if flags.Required {
@@ -157,7 +157,7 @@ func (tc *Collection) getSchemaByID(
 		}
 		return nil, err
 	}
-	schemaDesc, ok := desc.(catalog.SchemaDescriptor)
+	schemaDesc, ok := descs[0].(catalog.SchemaDescriptor)
 	if !ok {
 		return nil, sqlerrors.NewUndefinedSchemaError(fmt.Sprintf("[%d]", schemaID))
 	}

--- a/pkg/sql/catalog/descs/table.go
+++ b/pkg/sql/catalog/descs/table.go
@@ -63,7 +63,7 @@ func (tc *Collection) getTableByName(
 func (tc *Collection) GetLeasedImmutableTableByID(
 	ctx context.Context, txn *kv.Txn, tableID descpb.ID,
 ) (catalog.TableDescriptor, error) {
-	desc, _, err := tc.leased.getByID(ctx, tc.deadlineHolder(txn), tableID, false /* setTxnDeadline */)
+	desc, _, err := tc.leased.getByID(ctx, tc.deadlineHolder(txn), tableID)
 	if err != nil || desc == nil {
 		return nil, err
 	}
@@ -145,7 +145,7 @@ func (tc *Collection) GetImmutableTableByID(
 func (tc *Collection) getTableByID(
 	ctx context.Context, txn *kv.Txn, tableID descpb.ID, flags tree.ObjectLookupFlags,
 ) (catalog.TableDescriptor, error) {
-	desc, err := tc.getDescriptorByID(ctx, txn, tableID, flags.CommonLookupFlags)
+	descs, err := tc.getDescriptorsByID(ctx, txn, flags.CommonLookupFlags, tableID)
 	if err != nil {
 		if errors.Is(err, catalog.ErrDescriptorNotFound) {
 			return nil, sqlerrors.NewUndefinedRelationError(
@@ -153,7 +153,7 @@ func (tc *Collection) getTableByID(
 		}
 		return nil, err
 	}
-	table, ok := desc.(catalog.TableDescriptor)
+	table, ok := descs[0].(catalog.TableDescriptor)
 	if !ok {
 		return nil, sqlerrors.NewUndefinedRelationError(
 			&tree.TableRef{TableID: int64(tableID)})

--- a/pkg/sql/catalog/descs/type.go
+++ b/pkg/sql/catalog/descs/type.go
@@ -111,7 +111,7 @@ func (tc *Collection) GetImmutableTypeByID(
 func (tc *Collection) getTypeByID(
 	ctx context.Context, txn *kv.Txn, typeID descpb.ID, flags tree.ObjectLookupFlags,
 ) (catalog.TypeDescriptor, error) {
-	desc, err := tc.getDescriptorByID(ctx, txn, typeID, flags.CommonLookupFlags)
+	descs, err := tc.getDescriptorsByID(ctx, txn, flags.CommonLookupFlags, typeID)
 	if err != nil {
 		if errors.Is(err, catalog.ErrDescriptorNotFound) {
 			return nil, pgerror.Newf(
@@ -119,7 +119,7 @@ func (tc *Collection) getTypeByID(
 		}
 		return nil, err
 	}
-	switch t := desc.(type) {
+	switch t := descs[0].(type) {
 	case catalog.TypeDescriptor:
 		// User-defined type.
 		return t, nil

--- a/pkg/sql/reassign_owned_by.go
+++ b/pkg/sql/reassign_owned_by.go
@@ -169,7 +169,7 @@ func (n *reassignOwnedByNode) startExec(params runParams) error {
 func (n *reassignOwnedByNode) reassignDatabaseOwner(
 	dbDesc catalog.DatabaseDescriptor, params runParams,
 ) error {
-	mutableDbDesc, err := params.p.Descriptors().GetMutableDescriptorByID(params.ctx, dbDesc.GetID(), params.p.txn)
+	mutableDbDesc, err := params.p.Descriptors().GetMutableDescriptorByID(params.ctx, params.p.txn, dbDesc.GetID())
 	if err != nil {
 		return err
 	}
@@ -193,8 +193,7 @@ func (n *reassignOwnedByNode) reassignDatabaseOwner(
 func (n *reassignOwnedByNode) reassignSchemaOwner(
 	schemaDesc catalog.SchemaDescriptor, dbDesc *dbdesc.Mutable, params runParams,
 ) error {
-	mutableSchemaDesc, err := params.p.Descriptors().GetMutableDescriptorByID(
-		params.ctx, schemaDesc.GetID(), params.p.txn)
+	mutableSchemaDesc, err := params.p.Descriptors().GetMutableDescriptorByID(params.ctx, params.p.txn, schemaDesc.GetID())
 	if err != nil {
 		return err
 	}
@@ -218,8 +217,7 @@ func (n *reassignOwnedByNode) reassignSchemaOwner(
 func (n *reassignOwnedByNode) reassignTableOwner(
 	tbDesc catalog.TableDescriptor, params runParams,
 ) error {
-	mutableTbDesc, err := params.p.Descriptors().GetMutableDescriptorByID(
-		params.ctx, tbDesc.GetID(), params.p.txn)
+	mutableTbDesc, err := params.p.Descriptors().GetMutableDescriptorByID(params.ctx, params.p.txn, tbDesc.GetID())
 	if err != nil {
 		return err
 	}
@@ -248,8 +246,7 @@ func (n *reassignOwnedByNode) reassignTableOwner(
 func (n *reassignOwnedByNode) reassignTypeOwner(
 	typDesc catalog.TypeDescriptor, params runParams,
 ) error {
-	mutableTypDesc, err := params.p.Descriptors().GetMutableDescriptorByID(
-		params.ctx, typDesc.GetID(), params.p.txn)
+	mutableTypDesc, err := params.p.Descriptors().GetMutableDescriptorByID(params.ctx, params.p.txn, typDesc.GetID())
 	if err != nil {
 		return err
 	}

--- a/pkg/sql/repair.go
+++ b/pkg/sql/repair.go
@@ -694,7 +694,7 @@ func (p *planner) UnsafeDeleteDescriptor(ctx context.Context, descID int64, forc
 func unsafeReadDescriptor(
 	ctx context.Context, p *planner, id descpb.ID, force bool,
 ) (mut catalog.MutableDescriptor, notice error, err error) {
-	mut, err = p.Descriptors().GetMutableDescriptorByID(ctx, id, p.txn)
+	mut, err = p.Descriptors().GetMutableDescriptorByID(ctx, p.txn, id)
 	if mut != nil {
 		return mut, nil, nil
 	}

--- a/pkg/sql/resolver.go
+++ b/pkg/sql/resolver.go
@@ -610,7 +610,7 @@ func getDescriptorsFromTargetListForPrivilegeChange(
 					return nil, err
 				}
 				for i := range objectIDs {
-					descriptor, err := p.Descriptors().GetMutableDescriptorByID(ctx, objectIDs[i], p.txn)
+					descriptor, err := p.Descriptors().GetMutableDescriptorByID(ctx, p.txn, objectIDs[i])
 					if err != nil {
 						return nil, err
 					}
@@ -675,7 +675,7 @@ func getDescriptorsFromTargetListForPrivilegeChange(
 		}
 
 		for i := range objectIDs {
-			descriptor, err := p.Descriptors().GetMutableDescriptorByID(ctx, objectIDs[i], p.txn)
+			descriptor, err := p.Descriptors().GetMutableDescriptorByID(ctx, p.txn, objectIDs[i])
 			if err != nil {
 				return nil, err
 			}

--- a/pkg/sql/resolver.go
+++ b/pkg/sql/resolver.go
@@ -609,13 +609,13 @@ func getDescriptorsFromTargetListForPrivilegeChange(
 				if err != nil {
 					return nil, err
 				}
-				for i := range objectIDs {
-					descriptor, err := p.Descriptors().GetMutableDescriptorByID(ctx, p.txn, objectIDs[i])
-					if err != nil {
-						return nil, err
-					}
-					if descriptor != nil && descriptor.DescriptorType() == catalog.Table {
-						descs = append(descs, descriptor)
+				muts, err := p.Descriptors().GetMutableDescriptorsByID(ctx, p.txn, objectIDs...)
+				if err != nil {
+					return nil, err
+				}
+				for _, mut := range muts {
+					if mut != nil && mut.DescriptorType() == catalog.Table {
+						descs = append(descs, mut)
 					}
 				}
 			}
@@ -673,14 +673,13 @@ func getDescriptorsFromTargetListForPrivilegeChange(
 		if err != nil {
 			return nil, err
 		}
-
-		for i := range objectIDs {
-			descriptor, err := p.Descriptors().GetMutableDescriptorByID(ctx, p.txn, objectIDs[i])
-			if err != nil {
-				return nil, err
-			}
-			if descriptor != nil && descriptor.DescriptorType() == catalog.Table {
-				descs = append(descs, descriptor)
+		muts, err := p.Descriptors().GetMutableDescriptorsByID(ctx, p.txn, objectIDs...)
+		if err != nil {
+			return nil, err
+		}
+		for _, mut := range muts {
+			if mut != nil && mut.DescriptorType() == catalog.Table {
+				descs = append(descs, mut)
 			}
 		}
 	}

--- a/pkg/sql/schema_changer.go
+++ b/pkg/sql/schema_changer.go
@@ -434,7 +434,7 @@ func drainNamesForDescriptor(
 		}
 
 		// Free up the old name(s) for reuse.
-		mutDesc, err := descsCol.GetMutableDescriptorByID(ctx, descID, txn)
+		mutDesc, err := descsCol.GetMutableDescriptorByID(ctx, txn, descID)
 		if err != nil {
 			return err
 		}
@@ -453,7 +453,7 @@ func drainNamesForDescriptor(
 		// If the descriptor to drain is a schema, then we need to delete the
 		// draining names from the parent database's schema mapping.
 		if _, isSchema := mutDesc.(catalog.SchemaDescriptor); isSchema {
-			mutDB, err := descsCol.GetMutableDescriptorByID(ctx, mutDesc.GetParentID(), txn)
+			mutDB, err := descsCol.GetMutableDescriptorByID(ctx, txn, mutDesc.GetParentID())
 			if err != nil {
 				return err
 			}

--- a/pkg/sql/schemachanger/scdeps/exec_deps.go
+++ b/pkg/sql/schemachanger/scdeps/exec_deps.go
@@ -201,7 +201,7 @@ func (d *txnDeps) RemoveSyntheticDescriptor(id descpb.ID) {
 func (d *txnDeps) MustReadMutableDescriptor(
 	ctx context.Context, id descpb.ID,
 ) (catalog.MutableDescriptor, error) {
-	return d.descsCollection.GetMutableDescriptorByID(ctx, id, d.txn)
+	return d.descsCollection.GetMutableDescriptorByID(ctx, d.txn, id)
 }
 
 // NewCatalogChangeBatcher implements the scexec.Catalog interface.

